src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitTest.java

```diff
@@ -0,0 +1,111 @@
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+import static com.tngtech.archunit.library.dependencies.SlicesRuleDefinition.slices;
+import static java.nio.file.Files.lines;
+import static org.junit.Assert.assertEquals;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.difflib.DiffUtils;
+import com.github.difflib.UnifiedDiffUtils;
+import com.github.difflib.algorithm.DiffException;
+import com.github.difflib.patch.Patch;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.lang.ArchRule;
+
+public class ArchUnitTest {
+
+    /**
+     * Test contains asserts in callstack, but idea does not see them.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     */
+    @Test
+    public void packageCyclicDependencyTest() throws Exception {
+        final JavaClasses importedClasses = new ClassFileImporter()
+                .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+                .importPackages("com.puppycrawl.tools.checkstyle");
+
+        final ArchRule cyclicPackageRule = slices()
+                .matching("com.puppycrawl.tools.checkstyle.(**)..")
+            .should().beFreeOfCycles();
+        try {
+            cyclicPackageRule.check(importedClasses);
+        }
+        catch (AssertionError assertionError) {
+            processDiff(assertionError.getMessage());
+        }
+    }
+
+    /**
+     * Test contains asserts in callstack, but idea does not see them.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     */
+    @Test
+    public void nonProtectedCheckMethodsTest() {
+        final JavaClasses importedClasses = new ClassFileImporter()
+                .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+                .importPackages("com.puppycrawl.tools.checkstyle.checks");
+
+        final ArchRule checkMethodsShouldNotBeProtectedRule = methods().that()
+            .areDeclaredInClassesThat().haveNameMatching("^(Abstract).*Check")
+            .should().notBeProtected();
+
+        checkMethodsShouldNotBeProtectedRule.check(importedClasses);
+    }
+
+    private static void processDiff(String errorMsg) throws IOException, DiffException {
+        final List<String> prevSuppresions = lines(Paths.get("archunit_suppressions.txt"))
```

Comment:
please explain why https://www.archunit.org/userguide/html/000_Index.html#_ignoring_violations does not work for us.

Reply:
I used breakpoints in the archunit code corresponding to this feature.
The FailureMessages are of the type:
```
java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'slices matching 'com.puppycrawl.tools.checkstyle.(**)..' should be free of cycles' was violated (13 times):
Cycle detected: Slice api -> Slice checks.naming -> Slice api
Dependencies of Slice api
Method <com.puppycrawl.tools.checkstyle.api.AutomaticBean$RelaxedAccessModifierArrayConverter.convert(java.lang.Class, java.lang.Object)> calls method <com.puppycrawl.tools.checkstyle.checks.naming.AccessModifier.getInstance(java.lang.String)> in (AutomaticBean.java:398)
Dependencies of Slice checks.naming
Method 1....
Class 1....
Cycle detected: Slice api -> Slice checks.desgin -> Slice api
Method 1 ....
Method 2 ....
.
.
.
```

Here each section starting with `Cycle detected:` are treated as a single message in their filtering code.
What we aim to achieve is that, in `archunit_suppressions.txt` we should put
```
Method 1 ...
Method 2 ...
Class 1 ...
```
But in their `ignoring violation` feature if we specify `Method 1.*` as a regex whole of the message ie, the block starting with `Cycles detected: ` will be removed which is not desirable.
So, to treat individual `Method 1 ...`, `Method 2 ...` suppressions I have used `diff`.

Reply:
@romani I am facing a small issue here. Everytime I push the code(recently rebased on master), by the time the CI checks, there are new violations in the master and the check fails to pass.

Reply:
you did not answered my question, please do.
Please ask archunit maintainers (forum, stackverflow, .... ) on how to do suppression on all existing violations to prevent new violations to appear.

meanwhile please move to `config/archunit_suppressions.txt`.

Reply:
@romani the last answer explains why the `archunit_ignore_patterns.txt` do not work. I did ask the maintainers. Here's the link https://github.com/TNG/ArchUnit/issues/350
They also suggested to use freezing arch rule, the one which I implemented earlier.

Reply:
@romani Even if we use the feature in the link you mentioned(`archunit_ignore_patterns.txt`), we do need an UT to check the current violations and tell the user to either fix the new violation or that a exisiting violation has been fixed and needs to be removed from the suppression file, for which we require `diff`. If we use their feature we have to write all the violations in regex format which is a bit of extra work provided the way I have implemented already satisfies all needs. Please let me know if I am missing something.

Reply:
I try to avoid big files that keep state of some plugins.
Can you confirm that freezer will fail if any changes appear in store ? How diff on store will looks like ?

----

from referenced ticket in archunit ...
can we use `slices()...should().beFreeOfCycles().ignoreDependency` ?
and to make it useful ... we will create test method for each package, so our `ignoreXxxxxx` will be more specific. So we might be able to resolve dependency one by one and visually see what we still have in `ignoreXXXX`.
Lets make just another PR to try this.

Lets make one more PR with freeze rule, to see it in action on changes (second commit can remove some dependency by brutal code removal, some will fail but we need to see a diff on store file).

----

> So, to treat individual Method 1 ..., Method 2 ... suppressions I have used diff.

Why do we care that much of dependency of methods on class? I always thought of it as dependency between classes.
can we keep in repo archunit_ignore_patterns.txt as plain text file and do simple string compare by `google.truth` assert. If diff is happening, conributor just need  to generate new ignore file and send a change on it.